add set command

hqhq opened the pull request:
This PR add the new command set, which can setup or update configs of a container, right now it just used for set resource configs, such as memory, cpuset, etc.

hqhq commented:
Note that this PR need https://github.com/docker/libcontainer/pull/441 to me merged, so right now we might get compile error on Docker, please ignore that and review the code and docs first.

duglin commented:
Please squash your commits

hqhq commented:
@duglin  I took Michael's PR(https://github.com/docker/docker/pull/9984/commits) as example, daemon side, client side, docs, etc. Thought that's more clearer. Anyway, I squash the first three patches into one, I think the last one should be separated. Hope that's what you want.

hqhq commented:
ping, I don't know why Jenkins failed on set command, this is passed in my box.

hqhq commented:
ping @jfrazelle can you help take a look at Jenkins? Failed at `runCommand` seems quite strange.

cpuguy83 commented:
daemon/execdriver/native/driver.go:349: too many arguments in call to cont.Set

hqhq commented:
@cpuguy83 this is not the error for now, https://github.com/docker/libcontainer/pull/441 has been merged so arguments should be right. From currently Jenkins output, I don't know why it failed :(

cpuguy83 commented:
This needs a rebase now.

hqhq commented:
Rebased.

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,64 @@
+	for _, name := range names {
+		_, _, err := readBody(cli.call("POST", "/containers/"+name+"/set", hostConfig, false))
+		if err != nil {
+			fmt.Fprintf(cli.err, "%s\n", err)
```
Can you have this error elaborate a bit more, something like:
``` fmt.Fpritnf(cli.err, "Error trying to set properties of container (%s): %s\n", name, err )```

hqhq replied:
Updated and rebased, thanks.

duglin commented:
design LGTM

One question... I noticed that once you create a container with (or w/o) `--privileged` you can't change it.  Should we allow for that to change too? Can it?

hqhq commented:
@duglin I think even we want to do that, it can only be one way trip, once you drop a capability, you can never add it back before the system is rebooted, that's restriction from kernel.

jamtur01 commented on docs/man/docker-set.1.md:
```diff
@@ -0,0 +1,62 @@
+
+# DESCRIPTION
+
+Set resource configs for one or more containers. This can be the first
```
The second sentence needs a rewrite. It's not very clear.

hqhq replied:
Yes, it's not clear, and a bit redundant, I just get rid of it, do you have better suggestion?
Updated and rebased, thanks.

jamtur01 commented on docs/man/docker-set.1.md:
```diff
@@ -0,0 +1,62 @@
+
+Set resource configs for one or more containers. This can be the first
+time you setup the config which is not set when you create the container,
+or can be the update for configs which are already set. We can only set
```
As does this.

vishh commented:
+1. Thanks for working on this feature @hqhq!

tiborvass commented:
I don't have any objections for this feature, but it needs broad approval.

Ping @jfrazelle @crosbymichael @estesp @LK4D4 @icecrime @unclejack wdyt?

icecrime commented:
That's interesting, thanks! I like the idea, but the name `set` seems too generic and will make people wonder why `docker set --name foo` doesn't work.

Maybe that's a bit extreme, but if we accept this patch, we should perhaps agree today that `set` will have to support any (or at least *most*) of `run` parameters in the future. If we do, I think it would make sense to use the same `runconfig.Parse` function to parse the arguments, and have `set` error out on the provided settings which are not (yet) supported for dynamic update.

hqhq commented:
@icecrime Yes the name `set` is generic, the main problem I want to resolve is to support dynamic update resource configs, there is an approach before https://github.com/docker/docker/pull/3604, my approach was inspired from shykes's idea and took the name `set` he suggested. 

And yes, my design is support any (or at least most) of `run` parameters, so I think the name suits our needs, and that's confirmed from libcontainer side. But I think it's a bit early to use `runconfig.Parse` function, because I don't know if anyone would need to change `run` parameters besides resources, and that could make code a bit ugly. If we support more parameters of `run`, we can change to use `runconfig.Parse`, WDYT?

cpuguy83 commented:
We're already parsing resources for run and build... Maybe we can take this time to split the resource configs into a separate function and use in all three places.

hqhq commented:
@cpuguy83 Yes, that make sense, I did it in a separate PR https://github.com/docker/docker/pull/12235, please help review.

hqhq commented:
Hi, can we start reviewing this PR.

Updated:
 - rebased and remove engine usage
 - add support for new coming cgroup configs

cpuguy83 commented:
I'm definitely +1 on this design, ping @LK4D4 @crosbymichael @icecrime @tiborvass

calavera commented on daemon/set.go:
```diff
@@ -0,0 +1,57 @@
+	return warnings, nil
+}
+
+func (daemon *Daemon) updateResources(container *Container, hostConfig *runconfig.HostConfig) error {
```
there is nothing returning errors in this function, should we remove the `return` statement?

calavera commented:
I moved this PR to code-review since there are several maintainers that agree in the design. I have a minor comment but overall the code LGTM.

hqhq commented:
@calavera Thanks, updated.

hqhq committed:
Implemet docker set command

It's used for change configs of a container, we only support resource
configs for now, which is cgroup configs moved to hostconfig. It can
be extended in the future.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

duglin commented on daemon/set.go:
```diff
@@ -0,0 +1,53 @@
+		container.hostConfig.CpuQuota = hostConfig.CpuQuota
+	}
+
+	container.command.Resources.CpuShares = container.hostConfig.CpuShares
```
I didn't get a chance to look into it more, but I get a panic on this line because Resources is nil.

duglin replied:
ok I, by mistake, ran 'docker set' on a stopped container.  We need to add a check to make sure the container is running before we get to this point.

duglin replied:
hmmm now I'm not sure what happened, because I do see that you have a check for 'running' in the code.... still digging...

duglin replied:
weird.  When I run a container, let it stop, then try to do 'set' I get an error about it not running.  But, if I then reboot my daemon and run 'set' again against that container I get a panic.

duglin commented on daemon/set.go:
```diff
@@ -0,0 +1,53 @@
+	return warnings, nil
+}
+
+func (daemon *Daemon) updateResources(container *Container, hostConfig *runconfig.HostConfig) {
```
I would kind of prefer if the logic of updating the container properties were done by the container itself and not in the daemon.  IOW, have the daemon.updateResorces() & daemon.Set() calls above be a container.Set() (or 'Update') call instead.  Then we can add logging to that func too.

duglin replied:
We can also then check to see if the container is 'running' in container.Set() too.

hqhq replied:
Yes, seems reasonable, I'll change it, thanks.

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+func (cli *DockerCli) CmdSet(args ...string) error {
+	cmd := cli.Subcmd("set", "CONTAINER [CONTAINER...]", "Setup/Update resource configs for a container", true)
```
s/for a container/for a running container/

hqhq replied:
I'll redo this PR, and expend this feature to set a container wether it's running or stopped.

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,67 @@
+			}
+			MemorySwap = parsedMemorySwap
+		}
+	}
```
If the user doesn't specify any options what should we do?
- generate an error?
- call daemon but do nothing? current behavior
- just exit w/o calling daemon?

If nothing is going to change I don't see value in calling the daemon.  I don't have a strong opinion on this one, but I'm inclined to just exit (with zero exit code) w/o calling the daemon since technically its not an error, its just a no-op/pointless case.

hqhq replied:
What do you mean user doesn't specify any options? cmd require would require user specific at least one option, and ParseFlag would catch the wrong options. So I don't see how we would send daemon an empty calling. Correct me if I understand wrong.

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,67 @@
+		}
+	}
+
+	hostConfig := runconfig.HostConfig{
```
I don't think this is specific to this PR because I've been thinking about it for while but haven't found a situation where it actually manifested itself as an issue for an end-user until now, but the way we parse configs and then try to merge config structs is a problem.

If I run: `docker set -c 0 0685bc59c0dadfbb9971` the daemon has no way of knowing if zero means "the user specified zero" or "the user didn't specify it at all".  We really need to have the notion of sending a structure that has **just** the changed fields, no default values - then we know exactly what the user specified.

I would recommend that we:
- just send a json object with just the modified fields, no defaults
- have the daemon process each field individually instead of trying to deal with it as a single hostConfig struct

hqhq replied:
ok, I think your above comment is about this case? I considered this too and have discussion with other maintainer, current check with default value (0 or nil) to judge wether user set an option or not is not perfectly right, and haven't got any better idea to fix this.
Here I just think we should keep consistency with `docker run`, because options in feature would be a lot similar for these two.

duglin commented on daemon/set.go:
```diff
@@ -0,0 +1,53 @@
+		return warnings, err
+	}
+
+	container, err := daemon.Get(name)
```
Its a minor thing, but I'd prefer to check the name before the hostConfig data because to me that's the most important piece of information we get from the client so let's check that first.

hqhq replied:
Right, I'll check name first.

duglin commented on daemon/execdriver/native/driver.go:
```diff
@@ -355,6 +355,23 @@ func (d *driver) Stats(id string) (*execdriver.ResourceStats, error) {
 }
 
+func (d *driver) Set(c *execdriver.Command) error {
+	cont := d.activeContainers[c.ID]
```
ok, not 100% sure, but I think the panic issue I'm seeing is because you're checking to see if the container is 'active' instead of "running".

hqhq replied:
Let's see if the panic still exist when the new PR finished.

duglin commented on api/types/types.go:
```diff
@@ -24,6 +24,12 @@ type ContainerExecCreateResponse struct {
+// POST /containers/{name:.*}/set
+type ContainerSetResponse struct {
+	// Warnings are any warnings encountered during the execution of the command.
+	Warnings []string `json:"Warnings"`
```
They're really errors not warning, aren't they?

duglin commented on docs/man/docker-set.1.md:
```diff
@@ -0,0 +1,66 @@
+% Docker Community
+% JUNE 2014
+# NAME
+docker-set - Setup or update resource configs for one or more containers
```
s/containers/running containers/

duglin commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -2362,6 +2362,19 @@ more details on finding shared images from the command line.
+
+    Usage: docker set [OPTIONS] CONTAINER [CONTAINER...]
+
+    Set resource configs of one or more containers
```
s/containers/running containers/

duglin commented on docs/sources/reference/api/docker_remote_api_v1.19.md:
```diff
@@ -1054,6 +1054,44 @@ Status Codes:
+
+`POST /containers/(id)/set`
+
+Set resource configs of a container after it's created.
```
`Set resource config of a running container.`

duglin commented on docs/sources/reference/api/docker_remote_api_v1.19.md:
```diff
@@ -1054,6 +1054,44 @@ Status Codes:
+
+**Example request**:
+
+       POST /containers/create HTTP/1.1
```
create -> (id)/set

duglin commented on docker/flags.go:
```diff
@@ -107,6 +107,7 @@ func init() {
 			{"run", "Run a command in a new container"},
 			{"save", "Save an image to a tar archive"},
 			{"search", "Search for an image on the Docker Hub"},
+			{"set", "Setup or update resource configs for one or more containers"},
```
`Update properties of one or more running containers`.
If its too long, leave off "running".

icecrime commented:
*Collective review with @duglin @estesp @LK4D4 @jfrazelle @calavera*

Thank you for the PR, sorry for having been so long in giving it attention, and for the early review work.

After debate, we decided this is not a feature we should introduce:
- The use case is not general enough to be provided by Docker itself
- Where the use case is needed, it can already be achieved (for example with parent cgroup introduced in 1.6.0)
- Introducing a `SET` top-level command is an open door to many user confusion in the future (What can I set? Why can't I set X or Y? Why does it only work on running containers?).

We ultimately decided to close this.

hqhq commented:
@icecrime I don't get it.
 - The use case is definitely general, change container configs after it is created is no doubt useful, especially resource configs, I saw many people seek for this feature.
 - Sorry but I don't see any chance how our needs can be met with current features, for example with parent cgroup, can you be more specific?
 - This can be discussed more, but one thing for sure it wouldn't be limit to work on running containers only, just for the first step, this is easy to archive, we can surely expand that. And regarding which configs we can set, I don't think that'll be the problem, a simple `--help` would give users enough information which configs they can set. If someone wants to expand, let's just discuss if that's worth to do.

I don't know how you discussed this but I only see approval for the design in this thread from maintainers, and the reasons you give are really not strong enough. Hope you can reconsider it or give more reason why this feature should not be provided by Docker.

Thanks.

cpuguy83 commented:
Yes, I can see the usefulness of at the very least being able to tweak resource usages.
I thought this was something that @crosbymichael had specifically worked on getting into libcontainer so we could expose it in Docker... but I could be wrong.

Only thing I'm not happy about is that we have a `rename` endpoint and now this, but I digress.

jfrazelle commented:
I think what we are trying to avoid is the rabbit hole where then people
want everything to be supported by set, kinda like what happened with exec.
Today you can create cgroup-parents and easily change the resource
constraints.

Since this is really basic functionality we don't see why it can't be done
w a small tool that just changes the values of the resource files. That way
we don't have to go down the path where we have a new top level command and
have to support new features for everything it could do

On Friday, May 1, 2015, Brian Goff <notifications@github.com> wrote:

> Yes, I can see the usefulness of at the very least being able to tweak
> resource usages.
> I thought this was something that @crosbymichael
> <https://github.com/crosbymichael> had specifically worked on getting
> into libcontainer so we could expose it in Docker... but I could be wrong.
>
> Only thing I'm not happy about is that we have a rename endpoint and now
> this, but I digress.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11442#issuecomment-98126020>.
>

cpuguy83 commented:
What if this was API only? Maybe a post to /containers/<id>/resources?

duglin commented:
I'm not a fan of doing stuff like "API only" things.  I would interpret it as "we're going to support something but make it hard for you to get to it" (yea I know the API isn't hard, but its harder than the CLI).  If we're going to support something I think we should fully support it and make it a first-class feature and easy to use.

I think we should step back and not get bogged down in some of the technical side-issues. I think its important to first discuss, and decide, whether we want to support modifying an existing container's resource config properties via Docker?

IMO, if these configuration knobs are available in the underlying infrastructure then its hard for me to see why we wouldn't offer them to our users.

jfrazelle commented:
we can comment on this until the cows come home but my current belief is this is "feature creep" but that can be disproven in the future just like -f ;)

vishh commented:
`--cgroup-parent` should solve the update issue for cgroup based resources. The other option is to expose cgroup paths via inspect. 
Updating resources (and volumes maybe) is a feature that is very useful for managing clusters. 
One way to look at this would be to classify updates as power user features, and expect such power users to run additional software to handle updates. 
Is it fair to summarize that docker is not transforming into a system daemon for cluster management in the near future?

duglin commented:
my beloved -f  :-)

hqhq commented:
@jfrazelle One reason I use a more common command `set` is because someone tried a specific name https://github.com/docker/docker/pull/3604 and was told that we should not expose such low-level thing like cgroups to users, no mention to let users deal with cgroupfs directly. So I don't think `--cgroup-parent` can be a common solution.

Also, the design purpose including change other configs beside resources. I agree that people always have all kinds of needs for a common command, like we are keep adding options for `docker run`, that's something we should adjudge and maintain, if we can control `docker run`, we can control `docker set` in the same way. Even without `set`, if people want some feature, I won't be surprise if they propose other top command like `docker rename` to do one simple thing because they really have no other options.

Anyway, comments can be forever, I agree with @duglin , let's get to the same page about whether we want to support modifying an existing container's properties via Docker (not only resource configs).

cpuguy83 commented:
Is it fair to say that #6323 should be closed if we are not going to support this?

IMO since docker is the one setting up the cgroups why shouldn't it be the API we call to adjust them?
Nobody really used cgroups before docker (at least not the great masses that are today because of Docker) because who wants to have to do this?
Docker is the cgroup abstraction (among other things) I want to interact with.

duglin commented:
+1 to @cpuguy83   Docker is there to make people's lives easier so they don't need to get under the covers for everything the need to do. If the functionality is there in the underlying infrastructure its odd to me that we would want to hide it, unless exposing it breaks things or hacks things up.

hqhq commented:
@icecrime As we discussed in DockerCon, can we reopen this? Should we discussed how our approach should be or should I rebase first?
@crosbymichael Any comments on this PR?

hqhq commented:
Anyway I'll rebase it first, one change is that I would like to remove the restriction about `running` container, configs of any existed container should be changeable, maybe not paused container.
And I'll address all @duglin 's comments, thanks for these detailed review.

hqhq commented:
@icecrime @crosbymichael @duglin I have updated PR, can you please take a look.

I did forced update, but the code didn't change, is it because it's closed?

duglin commented:
@hqhq yes I believe github won't update it once the issue is closed.

ktraghavendra commented:
@hqhq 
Thanks for working on this. We definitely have a usecase for this. Immediate usecases I can think of is:
1. Suppose we start a database container with swappiness value x
if we observe that over a period of time we are seeing lot of major
faults reduce memory swappiness.

2. Container is cpu intensive and performance is low because of less
cpu-share given during container start.

ktraghavendra commented:
one unimportant thing: when I thought about this feature myself before looking at this thread,
I had something like below in mind. 
docker modify-config <image>  <config>=value

But I think docker set is more better and crispy.

selvik commented:
@icecrime  @jfrazelle -  Am trying to understand your suggestion of using --cgroup-parent to possbily change resource configs of a running container. Could you please provide an example to describe how this could be done ? 

The only way I can see to actually change the config. of a container that is already started is to directly write to the cgroup filesystem on the docker host.  

Thanks!

vishh commented:
@selvik: You will have to run a container with a `--cgroup-parent` path to be able to update resources. If the container is already running, that option won't work for you.

selvik commented:
@vishh Thanks a lot for the clarification. So this means that the proposed "docker set" command allows one to change resource configs of a *running* container but --cgroup parent cannot be used to do this.
The comments above seemed to indicate that docker set wasn't necessary because --cgroup-parent flag could help with the same..hence my confusion. Thanks!

zrml commented:
is the proposed SET command available then for a running contianer? 
I got lost trying to follow all the various links and variation of requests...
I think it's important for all management frameworks like Mesos, Kubernetes etc.
thank you

hqhq commented:
@zrml see https://github.com/docker/docker/pull/15078#issuecomment-136540061

the latest link is #15078
